modules/opus_multistream/opus_multistream.c

```diff
@@ -0,0 +1,209 @@
+/**
+ * @file opus_multistream.c Opus Multistream Audio Codec
+ *
+ * Copyright (C) 2019 Creytiv.com
+ */
+
+#include <re.h>
+#include <baresip.h>
+#include <opus/opus_multistream.h>
+#include "opus_multistream.h"
+
+
+/**
+ * @defgroup opus opus
+ *
+ * The OPUS multistream audio codec
+ *
+ * Supported version: libopus 1.0.0 or later
+ *
+ * Configuration options:
+ *
+ \verbatim
+  opus_stereo        yes     # Request peer to send stereo
+  opus_sprop_stereo  yes     # Sending stereo
+  opus_bitrate    128000     # Average bitrate in [bps]
+  opus_cbr        {yes,no}   # Constant Bitrate (inverse of VBR)
+  opus_inbandfec  {yes,no}   # Enable inband Forward Error Correction (FEC)
+  opus_dtx        {yes,no}   # Enable Discontinuous Transmission (DTX)
+  opus_complexity {0-10}     # Encoder's computational complexity (10 max)
+  opus_application {audio, voip} # Encoder's intended application
+ \endverbatim
+ *
+ * References:
+ *
+ *    RFC 6716  Definition of the Opus Audio Codec
+ *    RFC 7587  RTP Payload Format for the Opus Speech and Audio Codec
+ *
+ *    http://opus-codec.org/downloads/
+ */
+
+
+static bool opus_mirror;
+static char fmtp[256] = "";
+static char fmtp_mirror[256];
+
+uint32_t opus_complexity = 10;
+opus_int32 opus_application = OPUS_APPLICATION_AUDIO;
+
+uint32_t opus_ms_channels = 2;
+uint32_t opus_ms_streams = 2;
+uint32_t opus_ms_c_streams = 2;
+
+static int opus_multistream_fmtp_enc(struct mbuf *mb, const struct sdp_format *fmt,
+			 bool offer, void *arg)
+{
+	bool mirror;
+
+	(void)arg;
+	(void)offer;
+
+	if (!mb || !fmt)
+		return 0;
+
+	mirror = !offer && str_isset(fmtp_mirror);
+
+	return mbuf_printf(mb, "a=fmtp:%s %s\r\n",
+			   fmt->id, mirror ? fmtp_mirror : fmtp);
+}
+
+
+static struct aucodec opus_multistream = {
+	.name      = "opus_multistream",
+	.srate     = 48000,
+	.crate     = 48000,
+	.ch        = 2,
```

Comment:
this field says that the codec has 2 channels. I am wondering how this can work with 4 channels ..

Reply:
The `.ch` parameter of the` opus_multistream` struct gets updated from the default value ("2") in line 183 to the the `opus_ms_channels` value in the config file (e.g. "4").

Reply:
aha, I got it :)